File: mobile/lib/models/streak.dart
/// Learning-streak snapshot returned by `GET /streak`.
class Streak {
  final int currentStreak;
  final int longestStreak;
  final DateTime? lastLearningDay;
  final bool learnedToday;
  final int daysThisMonth;
  final int restoresUsed;
  final int restoresAvailable;
  final int maxRestoresPerMonth;
  final DateTime? restorableDay;
  final bool restoreEligible;

  const Streak({
    required this.currentStreak,
    required this.longestStreak,
    this.lastLearningDay,
    required this.learnedToday,
    required this.daysThisMonth,
    required this.restoresUsed,
    required this.restoresAvailable,
    required this.maxRestoresPerMonth,
    this.restorableDay,
    required this.restoreEligible,
  });

  factory Streak.fromJson(Map<String, dynamic> json) {
    return Streak(
      currentStreak: (json['current_streak'] as num?)?.toInt() ?? 0,
      longestStreak: (json['longest_streak'] as num?)?.toInt() ?? 0,
      lastLearningDay: json['last_learning_day'] != null
          ? DateTime.tryParse(json['last_learning_day'].toString())
          : null,
      learnedToday: json['learned_today'] as bool? ?? false,
      daysThisMonth: (json['days_this_month'] as num?)?.toInt() ?? 0,
      restoresUsed: (json['restores_used'] as num?)?.toInt() ?? 0,
      restoresAvailable: (json['restores_available'] as num?)?.toInt() ?? 0,
      maxRestoresPerMonth:
          (json['max_restores_per_month'] as num?)?.toInt() ?? 4,
      restorableDay: json['restorable_day'] != null
          ? DateTime.tryParse(json['restorable_day'].toString())
          : null,
      restoreEligible: json['restore_eligible'] as bool? ?? false,
    );
  }
}

File: frontend/src/api/streak.ts
// ─── streak.ts : learning-streak API ────────────────────────────────────
// Wrappers for the daily-streak endpoints: fetch current streak stats and
// restore a missed day using a monthly "freeze" budget.
import { apiRequest } from './client'

/** Streak summary returned by the backend. */
export type ApiStreak = {
  current_streak: number
  longest_streak: number
  last_learning_day: string | null
  learned_today: boolean
  days_this_month: number
  restores_used: number
  restores_available: number
  max_restores_per_month: number
  restorable_day: string | null
  restore_eligible: boolean
}

/** GET /api/streak — current streak, history, and restore availability. */
export function getStreak() {
  return apiRequest<ApiStreak>('/api/streak')
}

/** POST /api/streak/restore — revive the most recent missed learning day. */
export function restoreStreakDay() {
  return apiRequest<ApiStreak>('/api/streak/restore', {
    method: 'POST',
  })
}
